src/core/sentinel.d

```diff
@@ -0,0 +1,620 @@
+/**
+Contains types to differentiate arrays with sentinel values.
+*/
+module core.sentinel;
+
+/**
+Selects the default sentinel value for a type `T`.
+
+It has a special case for the char types, and also allows
+the type to define its own default sentinel value if it
+has the member `defaultSentinel`. Otherwise, it uses `T.init`.
+*/
+private template defaultSentinel(T)
+{
+         static if (is(Unqual!T ==  char)) enum defaultSentinel = '\0';
+    else static if (is(Unqual!T == wchar)) enum defaultSentinel = cast(wchar)'\0';
+    else static if (is(Unqual!T == dchar)) enum defaultSentinel = cast(dchar)'\0';
+    else static if (__traits(hasMember, T, "defaultSentinel")) enum defaultSentinel = T.defaultSentinel;
+    else                                   enum defaultSentinel = T.init;
+}
+
+// NOTE: T should be unqalified (not const/immutable etc)
+//       This "unqualification" of T is done by the `SentinelPtr` and `SentinelArray` templates.
+private template SentinelTemplate(T, immutable T sentinelValue)
+{
+    private enum CommonPtrMembers = q{
+        /**
+        Converts the ptr to an array by scanning it for the sentinel value to find its length.
+
+        Returns:
+            the ptr as a SentinelArray
+        */
+        inout(ArrayType) toArray() inout
+        {
+            ArrayType array = void;
+            array.array = cast(typeof(array.array))ptr[0 .. findLength()];
+            return cast(inout(ArrayType))array;
+        }
+
+        /**
+        Return the current value pointed to by `ptr`.
+        */
+        auto front() inout { return *ptr; }
+    };
+    struct MutablePtr
+    {
+        alias ArrayType = MutableArray;
+
+        union
+        {
+            T* ptr;
+            ConstPtr asConst;
+        }
+        alias asConst this; // facilitates implicit conversion to const type
+        // alias ptr this; // NEED MULTIPLE ALIAS THIS!!!
```

Comment:
https://github.com/dlang/dmd/pull/3998 though it will require quite an effort to revive this.